File: common/src/main/java/cn/breadnicecat/candycraftce/particle/CParticles.java
package cn.breadnicecat.candycraftce.particle;

import cn.breadnicecat.candycraftce.CandyCraftCE;
import dev.architectury.injectables.annotations.ExpectPlatform;
import net.minecraft.client.particle.ParticleEngine;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.core.particles.ParticleType;
import net.minecraft.core.particles.SimpleParticleType;
import net.minecraft.core.registries.BuiltInRegistries;

import java.util.function.Supplier;

import static cn.breadnicecat.candycraftce.utils.CommonUtils.impossibleCode;
import static cn.breadnicecat.candycraftce.utils.ResourceUtils.prefix;

/**
 * Created in 2024/4/5 下午5:50
 * Project: candycraftce
 *
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 * <p>
 */
public class CParticles {
	public static ParticleEntry<SimpleParticleType> CARAMEL_PORTAL_PARTICLE_TYPE = create("caramel_portal_particle", false, CaramelPortalParticle.CaramelProvider::new);
	
	private static ParticleEntry<SimpleParticleType> create(String name, boolean overrideLimiter, ParticleEngine.SpriteParticleRegistration<SimpleParticleType> provider) {
		ParticleEntry<SimpleParticleType> entry = register(name, () -> new SimpleParticleType(overrideLimiter));
		registerSprite(entry, provider);
		return entry;
	}
	
	@ExpectPlatform
	private static <T extends ParticleType<O>, O extends ParticleOptions> void registerSprite(ParticleEntry<T> type, ParticleEngine.SpriteParticleRegistration<O> provider) {
		impossibleCode();
	}
	
	private static <T extends ParticleType<?>> ParticleEntry<T> register(String name, Supplier<T> factory) {
		return CandyCraftCE.register(BuiltInRegistries.PARTICLE_TYPE, prefix(name), factory)
				.as(ParticleEntry::new);
	}
	
	public static void init() {
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/core/particle/CParticles.kt
package cn.breadnicecat.candycraftce.core.particle

import cn.breadnicecat.candycraftce.core.particle.particles.CaramelPortalType
import cn.breadnicecat.candycraftce.core.particle.particles.CaramelProvider
import cn.breadnicecat.candycraftce.utils.CLogUtils
import cn.breadnicecat.candycraftce.utils.CUtils.modLoc
import cn.breadnicecat.candycraftce.utils.CUtils.register
import cn.breadnicecat.candycraftce.utils.ifClient
import net.fabricmc.fabric.api.client.particle.v1.ParticleFactoryRegistry
import net.minecraft.core.particles.ParticleType
import net.minecraft.core.registries.BuiltInRegistries

/**
 * Created in 2024/4/5 下午5:50
 * Project: candycraftce
 *
 * @author [Bread_NiceCat](https://github.com/BreadNiceCat)
 *
 *
 */
object CParticles {
    init {
        CLogUtils.sign()
    }

    var caramel_portal_particle_type = register("caramel_portal_particle", CaramelPortalType())

    private fun <P : ParticleType<*>> register(id: String, type: P): P {
        return BuiltInRegistries.PARTICLE_TYPE.register(id.modLoc(), type)
    }

    init {
        ifClient {
            val instance = ParticleFactoryRegistry.getInstance()
            instance.register(caramel_portal_particle_type, ::CaramelProvider)
        }
    }

}
